### ISSUE TITLE
Rancher "Error connecting to WebSocket" on login when using rke2 as local management cluster

### ISSUE THREAD

**Issue by ShylajaDevadiga** [rancher/rancher#31431]
**Environment information**
RKE2 Version: rke2 v1.18.15+rke1r1, v1.19.7+rke2r1
Rancher Version: 2.5.5
NodeOS used to test: Ubuntu 20.04. CentOS 8.2, SELinux Disabled

**Cluster Configuration**
3 node cluster with etcd backend

**Describe the issue**
When using rke2 as local management cluster, Websocket error is seen during login

**Steps to reproduce (least amount of steps as possible):**

Rancher pods are in Running state.
```
 kubectl get nodes
NAME                                         STATUS   ROLES         AGE   VERSION
ip-172-31-3-73.us-east-2.compute.internal    Ready    etcd,master   70m   v1.19.7+rke2r1
ip-172-31-5-101.us-east-2.compute.internal   Ready    etcd,master   66m   v1.19.7+rke2r1
ip-172-31-9-203.us-east-2.compute.internal   Ready    etcd,master   67m   v1.19.7+rke2r1

kubectl get pods -n cattle-system
NAME                              READY   STATUS      RESTARTS   AGE
helm-operation-4mbws              0/2     Completed   0          48m
helm-operation-6p7d6              0/2     Completed   0          32m
helm-operation-7bmdv              0/2     Completed   0          49m
helm-operation-7jlwx              0/2     Completed   0          48m
helm-operation-g2d6d              0/2     Completed   0          48m
helm-operation-lg5c2              0/2     Completed   0          36m
helm-operation-n4nm9              0/2     Completed   0          48m
helm-operation-q2nrg              0/2     Completed   0          36m
helm-operation-v5xwt              0/2     Completed   0          49m
rancher-85c9cf85ff-9gx9w          1/1     Running     1          52m
rancher-85c9cf85ff-cjd5d          1/1     Running     0          52m
rancher-85c9cf85ff-rf6sk          1/1     Running     0          52m
rancher-webhook-b5b7b76c4-mr4sj   1/1     Running     0          48m
```
![image](https://user-images.githubusercontent.com/56045581/108799572-7b7e4c80-7545-11eb-8874-341791d3970a.png)

![image](https://user-images.githubusercontent.com/56045581/108803517-b4bbba00-754f-11eb-8326-ad53b169f90c.png)

Logs from the rancher pod:
```
W0223 03:58:04.312457       7 warnings.go:77] extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
2021/02/23 03:59:34 [WARNING] error updating ns user-bdfdg status: Operation cannot be fulfilled on namespaces "user-bdfdg": the object has been modified; please apply your changes to the latest version and try again
2021/02/23 03:59:41 [ERROR] error syncing 'fleet-system/helm-operation-6p7d6': handler helm-operation: an error on the server ("container not found (\"proxy\")") has prevented the request from succeeding (get pods helm-operation-6p7d6), requeuing
2021/02/23 03:59:41 [ERROR] error syncing 'fleet-system/helm-operation-6p7d6': handler helm-operation: an error on the server ("container not found (\"proxy\")") has prevented the request from succeeding (get pods helm-operation-6p7d6), requeuing
2021/02/23 03:59:41 [ERROR] error syncing 'fleet-system/helm-operation-6p7d6': handler helm-operation: an error on the server ("container not found (\"proxy\")") has prevented the request from succeeding (get pods helm-operation-6p7d6), requeuing
```

**Comment by bmdepesa** [comment]
@rancher-max @ShylajaDevadiga 

This issue is due to the setup of the environment. 

RKE2 deploys rke2-ingress-nginx-controller as a deployment instead of a daemonset. This setup is an HA install on 3 nodes fronted by a load balancer. When traffic from the LB is directed to the 2 nodes where the ingress controller is not running, the connection is refused. 

Followed the steps noted in the workaround here: https://github.com/rancher/rke2/issues/493 and the issue no longer occurs. 

Tested with: 

values.yaml
```
global:
  rke2DataDir: /var/lib/rancher/rke2
controller:
  kind: DaemonSet
  daemonset:
    useHostPort: true
  image:
    repository: us.gcr.io/k8s-artifacts-prod/ingress-nginx/controller
    tag: "v0.34.1"
```
```
helm upgrade rke2-ingress-nginx rke2/rke2-ingress-nginx -n kube-system -f values.yaml
```

Embedded kubectl & logs/exec shell also function correctly now (previously were flaky).

**Comment by davidnuzik** [comment]
Potentially related: https://github.com/rancher/rke2/issues/649

**Comment by davidnuzik** [comment]
Just to ensure we don't have something funky going on with support websockets with the default ingress helm chart config in rke2, and to ensure we are happy with our default RKE2 configuration here... @ShylajaDevadiga is going to set up an environment and share this with @brandond who will review it. We'll check and ensure it's working correctly as expected.

Then, if it's good to go this likely just ends up being a documentation task -- potentially in both rancher/rke2 and rancher/docs (RKE2 docs and Rancher2 docs).

**Comment by brandond** [comment]
@ShylajaDevadiga and I sat down and looked at how RKE2 is being deployed. In summary:
* The RKE2 cluster and AWS NLB are both being provisioned by Terraform. The ELB is statically configured to point at all nodes in the cluster; it does NOT use the AWS cloud controller to provision the ELB or manage which instances receive traffic.
* We experimented with enabling a Service of type=LoadBalancer for nginx ingress. The AWS cloud controller successfully created a new ELB, but it refuses to associate any of the nodes with the LB because they all have a `master` role, and the controller is hardcoded to ignore master nodes when selecting instances.

In summary, the best way to accommodate this pattern is for Terraform to drop a HelmChartConfig at `/var/lib/rancher/rke2/server/manifests/rke2-ingress-nginx-config.yaml` that tells the nginx ingress to use a Daemonset so that the ingress is available on all nodes, to match the NLB configuration that Terraform is creating.

```yaml
apiVersion: helm.cattle.io/v1
kind: HelmChartConfig
metadata:
  name: rke2-ingress-nginx
  namespace: kube-system
spec:
  valuesContent: |-
    controller:
      kind: DaemonSet
      daemonset:
        useHostPort: true
```

**Comment by davidnuzik** [comment]
We will approach documenting the necessary tweaks in rancher docs -- I have created an issue [here](https://github.com/rancher/docs/issues/3056)
This is a temporary solution. We will discuss this further and decide what next steps are needed later.

**Comment by bryson** [comment]
@brandond Attempting to create the ingress as a `DaemonSet` fails for me in a couple different ways:

> In summary, the best way to accommodate this pattern is for Terraform to drop a HelmChartConfig at `/var/lib/rancher/rke2/server/manifests/rke2-ingress-nginx-config.yaml` that tells the nginx ingress to use a Daemonset so that the ingress is available on all nodes, to match the NLB configuration that Terraform is creating.
> 
> ```yaml
> apiVersion: helm.cattle.io/v1
> kind: HelmChartConfig
> metadata:
>   name: rke2-ingress-nginx
>   namespace: kube-system
> spec:
>   valuesContent: |-
>     controller:
>       kind: DaemonSet
>       daemonset:
>         useHostPort: true
> ```

When Helm attempts to install Rancher 2 using the config quoted above, it fails a semantic version check:

```bash
$ kubectl logs -n kube-system helm-install-rke2-ingress-nginx-v6sl7
...
helm_v3 install --set-string global.rke2DataDir=/var/lib/rancher/rke2 --set-string global.systemDefaultRegistry=index.docker.io rke2-ingress-nginx /rke2-ingress-nginx.tgz --values /config/values-10_HelmChartConfig.yaml
Error: template: rke2-ingress-nginx/templates/controller-daemonset.yaml:80:22: executing "rke2-ingress-nginx/templates/controller-daemonset.yaml" at <semverCompare ">=0.9.0-beta.1" .Values.controller.image.tag>: error calling semverCompare: Invalid Semantic Version
```

I assume this is because the defaults from the chart [here](https://github.com/rancher/rke2-charts/blob/main-source/packages/rke2-ingress-nginx/generated-changes/patch/values.yaml.patch) include `tag: "nginx-0.35.0-rancher2"`.

When using those defaults directly in my `HelmChartConfig`, I get exactly the same error:

```yaml
apiVersion: helm.cattle.io/v1
kind: HelmChartConfig
metadata:
  name: rke2-ingress-nginx
  namespace: kube-system
spec:
  valuesContent: |-
    controller:
      kind: DaemonSet
      daemonset:
        useHostPort: true
      image:
        repository: rancher/nginx-ingress-controller
        tag: "nginx-0.35.0-rancher2"
```
```bash
$ kubectl logs -n kube-system helm-install-rke2-ingress-nginx-pzq7n
...
helm_v3 install --set-string global.rke2DataDir=/var/lib/rancher/rke2 --set-string global.systemDefaultRegistry=index.docker.io rke2-ingress-nginx /rke2-ingress-nginx.tgz --values /config/values-10_HelmChartConfig.yaml
Error: template: rke2-ingress-nginx/templates/controller-daemonset.yaml:80:22: executing "rke2-ingress-nginx/templates/controller-daemonset.yaml" at <semverCompare ">=0.9.0-beta.1" .Values.controller.image.tag>: error calling semverCompare: Invalid Semantic Version
```

Helm fails to install because the tag `nginx-0.35.0-rancher2` does not match semantic versioning standards with the `nginx-` prefix, and Helm refuses to accept the tag outright. As I understand it, unless the tag format of this image is changed in the `index.docker.io` repository itself, this will never work...unless there is an option to disable/bypass that semantic version check.

When changing the `repository` and `tag` values as instructed in the Rancher2 "HA on RKE2" documentation [here](https://rancher.com/docs/rancher/v2.5/en/installation/resources/k8s-tutorials/ha-rke2/) the Helm installation completes but the ingress pods fail to load:
```yaml
apiVersion: helm.cattle.io/v1
kind: HelmChartConfig
metadata:
  name: rke2-ingress-nginx
  namespace: kube-system
spec:
  valuesContent: |-
    controller:
      kind: DaemonSet
      daemonset:
        useHostPort: true
      image:
        repository: us.gcr.io/k8s-artifacts-prod/ingress-nginx/controller
        tag: "v0.34.1"
```
```bash
$ kubectl logs -n kube-system rke2-ingress-nginx-controller-5vhx7
Error from server (BadRequest): container "rke2-ingress-nginx-controller" in pod "rke2-ingress-nginx-controller-5vhx7" is waiting to start: trying and failing to pull image
```
```bash
$ kubectl describe pod -n kube-system rke2-ingress-nginx-controller-5vhx7
Name:         rke2-ingress-nginx-controller-5vhx7
Namespace:    kube-system
...
Controlled By:  DaemonSet/rke2-ingress-nginx-controller
Containers:
  rke2-ingress-nginx-controller:
    Container ID:
    Image:         index.docker.io/us.gcr.io/k8s-artifacts-prod/ingress-nginx/controller:v0.34.1
 ...
Events:
  Type     Reason     Age                     From               Message
  ----     ------     ----                    ----               -------
  Normal   Scheduled  8m18s                   default-scheduler  Successfully assigned kube-system/rke2-ingress-nginx-controller-5vhx7 to ip-10-0-11-118
  Normal   Pulling    6m40s (x4 over 8m17s)   kubelet            Pulling image "index.docker.io/us.gcr.io/k8s-artifacts-prod/ingress-nginx/controller:v0.34.1"
  Warning  Failed     6m38s (x4 over 8m16s)   kubelet            Failed to pull image "index.docker.io/us.gcr.io/k8s-artifacts-prod/ingress-nginx/controller:v0.34.1": rpc error: code = Unknown desc = failed to pull and unpack image "docker.io/us.gcr.io/k8s-artifacts-prod/ingress-nginx/controller:v0.34.1": failed to resolve reference "docker.io/us.gcr.io/k8s-artifacts-prod/ingress-nginx/controller:v0.34.1": pull access denied, repository does not exist or may require authorization: server message: insufficient_scope: authorization failed
  Warning  Failed     6m38s (x4 over 8m16s)   kubelet            Error: ErrImagePull
  Normal   BackOff    6m12s (x6 over 8m15s)   kubelet            Back-off pulling image "index.docker.io/us.gcr.io/k8s-artifacts-prod/ingress-nginx/controller:v0.34.1"
  Warning  Failed     3m16s (x19 over 8m15s)  kubelet            Error: ImagePullBackOff
```

The problem here is that the image that the containers are trying to pull is _not_ the image from the YAML config, it is that image with the string `index.docker.io/` prepended, as you can see in the events list: 

```
Failed to pull image "index.docker.io/us.gcr.io/k8s-artifacts-prod/ingress-nginx/controller:v0.34.1": rpc error: code = Unknown desc = failed to pull and unpack image "docker.io/us.gcr.io/k8s-artifacts-prod/ingress-nginx/controller:v0.34.1": failed to resolve reference "docker.io/us.gcr.io/k8s-artifacts-prod/ingress-nginx/controller:v0.34.1": pull access denied, repository does not exist or may require authorization: server message: insufficient_scope: authorization failed
```

I notice in the first failure using your quoted config that Helm's install command includes `--set-string global.systemDefaultRegistry=index.docker.io`. Not sure that this is related as it's just a default, and I wouldn't expect the default to override the value given...

As far as I can tell, there is no "correct" way to create this DaemonSet in an automated way with this combination of bugs. Is there another solution? 

Currently as a work-around I use the YAML config from the Rancher docs so Helm's install "completes" and the Pods are created/scheduled but image pull fails, and I then set the DaemonSet image manually using Terraform's `depends_on` directives to create a bottleneck for everything that happens after the cluster is expected to be ready:

```
resource "null_resource" "kubernetes_cluster_init" {
  provisioner "local-exec" {
    interpreter = ["bash", "-c"]
    command     = "until kubectl set image -n kube-system ds/rke2-ingress-nginx-controller rke2-ingress-nginx-controller=rancher/nginx-ingress-controller:nginx-0.35.0-rancher2; do sleep 5; done"
  }

  depends_on = [
    # ... (RKE2 server node creation, load balancer availability, dns, etc)
    null_resource.get_kubeconfig,
  ]
}

resource "helm_release" "cert_manager" {
  # ...
  depends_on = [null_resource.kubernetes_cluster_init]
}
resource "helm_release" "rancher2" {
  repository = "https://releases.rancher.com/server-charts/latest"
  # ...
  depends_on = [helm_release.cert_manager]
}
```

It seems that these bugs affect multiple repositories/projects, so posting here since this discussion is ongoing.

Helm: `v3.5.3`
Rancher: `v2.5.7`
RKE2: `v1.20.4+rke2r1`
EC2 instances: `ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-20201014` (`ami-84556de5` in Gov Cloud)
Terraform:
```
Terraform v0.14.8
+ provider registry.terraform.io/cloudflare/cloudflare v2.19.2
+ provider registry.terraform.io/hashicorp/aws v3.32.0
+ provider registry.terraform.io/hashicorp/helm v2.0.3
+ provider registry.terraform.io/hashicorp/kubernetes v2.0.2
+ provider registry.terraform.io/hashicorp/null v3.1.0
+ provider registry.terraform.io/hashicorp/random v3.1.0
+ provider registry.terraform.io/rancher/rancher2 v1.12.0
```

**Comment by bryson** [comment]
@brandond the `HelmChartConfig` that you posted does not work. I've created a detailed bug report at rancher/rke2#493.

**Comment by brandond** [comment]
@bryson that fix is probably specific to the AWS deployment automation that our QA team uses. It was not meant to be a general recommendation for all users.

**Comment by bryson** [comment]
> @bryson that fix is probably specific to the AWS deployment automation that our QA team uses: I believe it works if you have it there during the initial install. It was not meant to be a general recommendation for all users.

Are you referring to the YAML you recommended for the HelmChartConfig? It's almost identical to the Rancher documentation's recommendation [here](https://rancher.com/docs/rancher/v2.5/en/installation/resources/k8s-tutorials/ha-rke2/#5-configure-nginx-to-be-a-daemonset).

**Comment by davidnuzik** [comment]
@ShylajaDevadiga as a low priority (lower than today's releases) can we look into this again as per the comments? We should re-test and ensure RKE2 is configured correctly for Rancher. If get stuck, raise with Max and I and perhaps we can sort it out together. Does not appear to be a Rancher-specific issue.

**Comment by sangeethah** [comment]
Waiting for  rancher/rke2#493 to be addressed.

**Comment by ShylajaDevadiga** [comment]
Validated fix https://github.com/rancher/rke2/issues/493 using rke2 commit id
rke2 version v1.20.5-dev+c7da840c
rancher version 2.5.7

Placed HelmChartConfig at /var/lib/rancher/rke2/server/manifests/rke2-ingress-nginx-config.yaml, ingress-controller is now deployed as a daemonset and available in all nodes.
```
cat rke2-ingress-nginx-config.yaml 
apiVersion: helm.cattle.io/v1
kind: HelmChartConfig
metadata:
  name: rke2-ingress-nginx
  namespace: kube-system
spec:
  valuesContent: |-
    controller:
      kind: DaemonSet
      daemonset:
        useHostPort: true
```

**Comment by izaac** [comment]
This only works on rke2 1.20+

I tried today with 


`v1.18.18+rke2r1`

And got:

```
Error: UPGRADE FAILED: template: rke2-ingress-nginx/templates/controller-daemonset.yaml:80:22: executing "rke2-ingress-nginx/templates/controller-daemonset.yaml" at <semverCompare ">=0.9.0-beta.1" .Values.controller.image.tag>: error calling semverCompare: Invalid Semantic Version
```

And similar result with  `v1.19.10+rke2r1`

**Comment by brandond** [comment]
Yes, as noted in the milestone on the linked rke2 issue.
